### ISSUE TITLE
Sktime metrics rework

### ISSUE THREAD

**Issue by chrisholder** [alan-turing-institute/sktime#1062]
**Is your feature request related to a problem? Please describe.**
When compared to the vast metric library that sklearn has, sktime is missing a great deal of structure. Additionally some of the metrics in Sktime are inconsistent with sklearn. See: https://scikit-learn.org/stable/modules/classes.html#module-sklearn.metrics for details on what sklearn has to offer.

In particular as someone working on the clustering side of sktime the metrics are imperative to many of the algorithms for both funcationality and offering evaluation. 

If you consider the current directory of sktime we have distances, benchmarking, performance_metrics in the upper most level, then some plotting stuff in with the utils, some evaluation methods within a type of time series analysis (i.e. forecasting has model_evluation) and some like classifiction where it isn't obvius where the evaluation metrics are. Point being while this is working for now the inconsistency and departure from the structure of sklearn is confusing for both developers and users of the package.

**Describe the solution you'd like**
My proposal is a restructure of sktime so that model selection, classification metrics, regression metrics, forecasting metrics, clustering metrics, pairwise metrics (distances) and plotting are located under the sktime/metrics/ directory which is in line with sklearn. In addition to the restructure I would propose some parts (in particular distance metrics) are reworked and integrated with numba rather than cython for future ease of development (see http://numba.pydata.org/).

In particular I personally would like to prioritise working on the pairwise distance. Currently the distances I'm using for clustering  are located in distances/elastic.py and distances/elastic_cython.pyx. While these distance work and are very well optimised,  they don't follow the interface set out by sklearn (they do not allow for 2 1d arrays to be past). While im not propsing scrapping these overnight I believe overtime they should be redeveloped to follow sklearns interface and move to numba over Cython.

**Describe alternatives you've considered**
Alternativly without this structure I would just develop the distance and evaluation methods I need inside the sktime/clustering directory which further makes the structure of sktime confusing and inconsistent.

**Additional context**
This is a large proposal and restructure. However, I think if we do it incrementally we can get to a good point of consistency. In particular while talking with @TonyBagnall the distance metrics we agreed that accepting 1d arrays is a necessity for Sktime.

**Comment by TonyBagnall** [comment]
Whilst I largely agree about performance measures being grouped, I think your are conflating performance  metrics with distances/metrics, and this is really two issues
1. Redesign distances with numba. I think this would be great if it does not degrade performance, but I think it makes sense to keep distances a separate module
2. Group performance/evaluating code in one place. This includes the benchmarking module

**Comment by chrisholder** [comment]
I agree they should be seperated. The main reason I suggested them being grouped together is simply sklearn does it like that. The sklearn documentations states:

> The sklearn.metrics module includes score functions, performance metrics and pairwise metrics and distance computations.

However, another package such as Tslearn does it in the way @TonyBagnall describes where they have distance metrics in a seperate directory to becnhmarking and scoring. 

I overall think the structure @TonyBagnall suggested is better and the two issues he outlined should be created.

**Comment by TonyBagnall** [comment]
ah I see, well I'm not in favour of slavishly following sklearn, I think they should be separate. However, there is a trade off as you say. We should discuss this on the core contributor channel, I will nominate you for inclusion when the clustering PR goes in

**Comment by fkiraly** [comment]
agreed that metrics should be collected, and agreed with @TonyBagnall that metrics and distances should be separated.

I have some code for distances, would contribute that soon-ish.

>ah I see, well I'm not in favour of slavishly following sklearn

Oh no 😱, how you've said it, @TonyBagnall! You'll bring eternal misfortune upon us.

**Comment by RNKuhns** [comment]
I also like the idea of collecting the performance metrics into a single area (distance functions aside). Would the plan be to add them under performance_metrics or rename as metrics (with appropriate deprecation) with performance_metrics as a potential sub-module?

**Comment by fkiraly** [comment]
Either looks sensible to me, though I also don't see a strong reason for renaming.

**Comment by fkiraly** [comment]
what metrics would be that are not performance metrics?

If you think about distance metrics, I'd consider them as different.

Though I admit that there is not much conceptual distance between a distance metric and a performance metric. Where, of course, not all distances are metrics in the math sense.

**Comment by fkiraly** [comment]
completed here for performance metrics:
https://github.com/alan-turing-institute/sktime/pull/2502

and via the dists_kernels interface.
